File: hoiiday-front/hoiiday-frontend/src/app/services/property.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Property } from '../model/property/property.model';
import { PropertyInsertDTO } from '../model/property/property-insert.model';

export interface PropertyWithPrice extends Property {
  lowestPrice: number;
}

@Injectable({
  providedIn: 'root',
})
export class PropertyService {
  private apiUrl = 'http://localhost:8080/api/properties';
  private readonly BASE = '/api/properties';
  constructor(private http: HttpClient) {}

  getAllProperties(): Observable<Property[]> {
    return this.http.get<Property[]>(this.BASE);
  }
  
  getPropertyById(id: number): Observable<Property> {
    const url = `${this.BASE}/${id}`;
    return this.http.get<Property>(url);
  }

  createProperty(property: PropertyInsertDTO): Observable<Property> {
    return this.http.post<Property>(this.BASE, property);
  }

  updateProperty(property: Property): Observable<Property> {
    const url = `${this.BASE}/${property.propertyId}`;
    return this.http.put<Property>(url, property);
  }

  deleteProperty(id: number): Observable<void> {
    const url = `${this.BASE}/${id}`;
    return this.http.delete<void>(url);
  }
  
}

File: hoiiday-backend/src/main/java/com/example/hoiiday/service/PropertyService.java
package com.example.hoiiday.service;

import com.example.hoiiday.DTO.PropertyDTO;
import com.example.hoiiday.DTO.PropertyInsertDTO;

import java.math.BigDecimal;
import java.util.List;

public interface PropertyService {
    List<PropertyDTO> getAllProperties();
    PropertyDTO getPropertyById(Long id);
    PropertyDTO createProperty(PropertyInsertDTO propertyInsertDTO);
    PropertyDTO updateProperty(Long id, PropertyDTO propertyDTO);
    void deleteProperty(Long id);
    BigDecimal getLowestPriceForProperty(Long propertyId);
}
